consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentReservoirSamplingSpanProcessor.java

```diff
@@ -0,0 +1,564 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.contrib.samplers;
+
+import static io.opentelemetry.api.internal.Utils.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+import io.opentelemetry.api.trace.SpanContext;
+import io.opentelemetry.api.trace.TraceState;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.sdk.common.CompletableResultCode;
+import io.opentelemetry.sdk.internal.DaemonThreadFactory;
+import io.opentelemetry.sdk.trace.ReadWriteSpan;
+import io.opentelemetry.sdk.trace.ReadableSpan;
+import io.opentelemetry.sdk.trace.SpanProcessor;
+import io.opentelemetry.sdk.trace.data.DelegatingSpanData;
+import io.opentelemetry.sdk.trace.data.SpanData;
+import io.opentelemetry.sdk.trace.export.SpanExporter;
+import java.util.ArrayList;
+import java.util.BitSet;
+import java.util.Collections;
+import java.util.List;
+import java.util.PriorityQueue;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+/**
+ * A {@link SpanProcessor} which periodically exports a fixed maximum number of spans. If the number
+ * of spans in a period exceeds the fixed reservoir (buffer) size, spans will be consistently
+ * (compare {@link ConsistentSampler}) sampled.
+ */
+public final class ConsistentReservoirSamplingSpanProcessor implements SpanProcessor {
+
+  private static final String WORKER_THREAD_NAME =
+      ConsistentReservoirSamplingSpanProcessor.class.getSimpleName() + "_WorkerThread";
+
+  private final Worker worker;
+  private final AtomicBoolean isShutdown = new AtomicBoolean(false);
+
+  public static final long DEFAULT_EXPORT_TIMEOUT_NANOS = TimeUnit.SECONDS.toNanos(30);
+
+  private static final class ReadableSpanWithPriority {
+
+    private final ReadableSpan readableSpan;
+    private int pval;
+    private final int rval;
+    private final long priority;
+
+    public static ReadableSpanWithPriority create(
+        ReadableSpan readableSpan, RandomGenerator randomGenerator) {
+      String otelTraceStateString =
+          readableSpan.getSpanContext().getTraceState().get(OtelTraceState.TRACE_STATE_KEY);
+      OtelTraceState otelTraceState = OtelTraceState.parse(otelTraceStateString);
+      int pval;
+      int rval;
+      long priority = randomGenerator.nextLong();
```

Comment:
I understand that among spans with the same r-value, the **priority** decides which spans will be removed from the reservoir. Giving each span a new random value may break trace completeness to a larger degree than it is necessary.
To alleviate this problem, the priority might be made dependent on trace-id, for example by combining a hash of trace-id and a (less significant) random number. Since there are no guarantees about quality of trace-id in general, I do not advocate to do it by default, but it would be nice to refactor the code so it would be easy to do for the users who really want it.

Reply:
Thanks, this improvement could be added in a later PR without breaking anything. As it would require choosing a hash function, which is a difficult decision by itself, much more discussion would be needed. (Ideally, it should be a fast, high-quality function for which there are compatible implementations in all relevant programming languages.) I think as a first step, this PR should be merged without this improvement.

Reply:
What I meant by refactoring was something like calculating the priority using an overridable method with trace-id as a parameter. It would remain unused in the default implementation.
But I agree with you that it can be done later.